src/cart/test/utest/utest_portnumber.c

```diff
@@ -0,0 +1,186 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. 8F-30005.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * This file is part of CaRT testing.
+ */
+
+/*
+ * This code tests whether a provider returns an error if 2 independent
+ * instances attempts to open the same port number.
+ *
+ * Two child process are forked with the same provider information.
+ * The test is setup so that first child opens the port and then
+ * the second child should fail.
+ * Note, synchronization between child process is performed via sleep
+ * function calls.  The sleep time should prevent any problems, but
+ * beware if an issue arise where the results are swapped.
+ */
+/*
+ * Reference jira ticket DAOS-5732 to include socket and verb tests.
+ */
+#define tests_not_included
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <time.h>
+#include <sys/wait.h>
+
+#include <cmocka.h>
+#include <cart/api.h>
+#include "../cart/crt_internal.h"
+
+
+static void
+run_test_fork(void **state)
+{
+	int	result1 = -1;
+	int	result2 = -1;
+	int	status = -1;
+	int	rc = 0;
+	int	child_result;
+	pid_t	pid1 = 0;
+	pid_t	pid2 = 0;
+	crt_context_t crt_context = NULL;
+
+	pid1 = fork();
+	/* fork first child process */
+	if (pid1 == 0) {
+		rc = crt_init(NULL, CRT_FLAG_BIT_SERVER);
+		if (rc != 0)
+			exit(10);
+		child_result = crt_context_create(&crt_context);
+		sleep(10);	/* wait for second child */
+		rc = crt_context_destroy(crt_context, false);
+		if (rc != 0)
+			exit(11);
+		rc = crt_finalize();
+		if (rc != 0)
+			exit(12);
+		exit(child_result);
+	}
+
+	/* fork second child process */
+	pid2 = fork();
+	if (pid2 == 0) {
+		sleep(2);  /* wait for first child just in case */
+		rc = crt_init(NULL, CRT_FLAG_BIT_SERVER);
+		if (rc != 0)
+			exit(20);
+		child_result = crt_context_create(&crt_context);
+		rc = crt_context_destroy(crt_context, false);
+		if (rc != 0)
+			exit(21);
+		rc = crt_finalize();
+		if (rc != 0)
+			exit(22);
+		exit(child_result);
+	}
+
+	/* Wait for first child and get results */
+	waitpid(pid1, &status, 0);
+	if (WIFEXITED(status)) {
+		result1 = WEXITSTATUS(status);
+	}
+
+	/* Wait for second child and get results */
+	waitpid(pid2, &status, 0);
+	if (WIFEXITED(status)) {
+		result2 = WEXITSTATUS(status);
+	}
+
+	/* Test results.  first child should should succeed. */
+	assert_true(result1 == 0);
+	assert_true(result2 != 0);
```

Comment:
here we want to also assert_true(result2 != 20, 21, 22).
We only want test to pass if crt_context_create() failed, but if anything else failed in child2 prior to crt_context_create() then we should fail the test.

Also in your test above on Line 94 crt_context_destroy() will likely fail (since we failed to create a context in a first place), so it seems you might be getting back error 21 here for second child all the time (code will never reach Line 100)

Reply:
Add assert test "assert_false" for case 20, 21 and 22.
Modified child 2 code to call crt_context_destroy only if the create succeed (the socket failure case).  Note the for tcp, crt_context_destroy does not return a failure even though the create fails.